### PR TITLE
Mukuna solution 2

### PR DESCRIPTION
**Pull request by kerthnorth**
---
name: Triangle Printer 
about: A problem that prints out triangle based on a given row
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: .vscode/settings.json
Comment (RaedAlyan) [inline .vscode/settings.json:1]: pls don't push this file

File: solutions/print_triangle_shape.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+"""
+This function prints a triangle shape made of asterisks ("*").
+
+Module: print_triangle_shape
+
+Functions:
+    triangle_printer(user_input):
+
+@Author: Mukuna Kabeya
+@Date: 2024-01-06
```
Comment (RaedAlyan) [inline solutions/print_triangle_shape.py:10]: please edit the format of the date

File: solutions/print_triangle_shape.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        A string representing the asterisks for each row of the triangle. The number of
+        asterisks follows an odd-numbered sequence (1, 3, 5, 7, etc.), calculated as
+        `2 * i - 1`, where `i` is the current row index.
+
+    holder : list
+        A list that stores each row of the triangle as a string. Each row contains
+        a combination of spaces (`s`) and asterisks (`x`) that form the desired shape.
+        This list is returned after the triangle is printed.
+    """
+    if not str(user_input).isdigit():
```
Comment (RaedAlyan) [inline solutions/print_triangle_shape.py:75]: if you want to check if the number is int or not, you can do it like this: assert isinstance(user_input, int)

File: solutions/print_triangle_shape.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        raise TypeError("Please enter an integer only!")
+    rows = int(user_input)
```
Comment (RaedAlyan) [inline solutions/print_triangle_shape.py:77]: if you want the user_input be an integer, why did you pass it as a str?

File: solutions/print_triangle_shape.py (hunk 2 above)
Comment (kerthnorth) [reply solutions/print_triangle_shape.py:75]: Hi Raed,
Thank you for your suggestion!

As per TDD principles, the code was written after the test cases, which currently check for type and value errors. Adapting the tests to fit this method would require changes that may not align with the core practices of TDD.

That said, your solution offers a simplified approach, and I truly appreciate it. I will definitely consider using it in the future where appropriate.

Thank you again for sharing this!

File: solutions/print_triangle_shape.py (hunk 3 above)
Comment (kerthnorth) [reply solutions/print_triangle_shape.py:77]: Great find. I will just pass rows as user_input. 

Thank you for highlighting this!

File: solutions/print_triangle_shape.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if rows <= 1:
+        raise ValueError("Please provide an integer greater than 1.")
+    holder = []
+    for i in range(1, rows + 1):
+        s = " " * (rows - i)
+        x = "*" * (2 * i - 1)
+        holder.append(s + x)
+    # print("\n".join(holder))
```
[the remark this answers is not in the document]
Comment (kerthnorth) [reply solutions/print_triangle_shape.py:85]: Thank You for notifying me of this. I'll attend to it!

**Review by RaedAlyan — APPROVED** [review]
This PR look fine to me
